tests/first_stage_test.py

```diff
@@ -16,37 +20,322 @@ class CommandLineTest(testlib.RouterMixin, testlib.TestCase):
+            stdin=subprocess.PIPE,
+        )
+        try:
+            try:
+                stdout, stderr = proc.communicate(input=self.preamble, timeout=10)
+            except TypeError:
+                stdout, stderr = proc.communicate(input=self.preamble)
+        except:
+            proc.kill()
+            self.fail("First stage did not finish")
+        else:
+            # proc was killed by SIGTERM?!? TODO Where does this come from?
+            self.assertEqual(-15, proc.returncode)
+            self.assertEqual(
+                stdout,
+                mitogen.parent.BootstrapProtocol.EC0_MARKER
+                + b("\n")
+                + mitogen.parent.BootstrapProtocol.EC1_MARKER
+                + b("\n")
+                + mitogen.parent.BootstrapProtocol.EC2_MARKER
+                + b("\n"),
+            )
+            self.assertEqual(
+                b(""),
+                stderr,
+            )
+
+    def test_stdin_non_blocking(self):
+        """Test that first stage works with non-blocking STDIN
+
+        The boot command should read the preamble from STDIN, write all ECO
+        markers to STDOUT, and then execute the preamble.
+
+        This test writes the complete preamble to non-blocking STDIN.
+
+        1. Fork child reads from non-blocking STDIN
+        2. Fork child writes all ECO markers to stdout as expected.
+
+        """
+
+        PIPE_SIZE = 4096
+        # Make sure that not all data can be written in one write operation.
+        CHUNK_SIZE = 2 * PIPE_SIZE
+        r, w = mitogen.core.pipe(blocking=False)
+        with w:
+            try:
+                fcntl.fcntl(w.fileno(), fcntl.F_SETPIPE_SZ, PIPE_SIZE)
+            except AttributeError:
+                pass
+            else:
+                self.assertEqual(fcntl.fcntl(w.fileno(), fcntl.F_GETPIPE_SZ), PIPE_SIZE)
+
+            proc = subprocess.Popen(
+                args=self.args,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                stdin=r,
+                close_fds=True,
+            )
+            # Close the read fd in the parent
+            r.close()
+
+            view = mitogen.core.BufferType(self.preamble, 0)
+            offset = 0
+            while offset < len(view):
+                end = min(offset + CHUNK_SIZE, len(view))
+                _, w_fds, _ = select.select([], [w.fileno()], [], 10)
+                self.assertIn(w.fileno(), w_fds)
+                written = os.write(w.fileno(), view[offset:end])
+                self.assertGreater(written, 0)
+                offset += written
+
+        try:
+            try:
+                returncode = proc.wait(timeout=10)
+            except TypeError:
+                returncode = proc.wait()
+        except:
+            proc.kill()
+            self.fail("First stage did not finish")
+        else:
+            try:
+                # proc was killed by SIGTERM?!? TODO Where does this come from?
```

Comment:
Probably similar or same cause as #1360

Reply:
No, I figured out the reason: the broker killed itself intentionally.